internal/stack/stacks.go

```diff
@@ -67,45 +69,88 @@ func (s Stack) String() string {
 }
 
 func getStacks(all bool) []Stack {
-	var stacks []Stack
+	stacks, err := newStackParser(bytes.NewReader(getStackBuffer(all))).Parse()
+	if err != nil {
+		panic(err)
+	}
+	return stacks
+}
 
-	var curStack *Stack
-	stackReader := bufio.NewReader(bytes.NewReader(getStackBuffer(all)))
-	for {
-		line, err := stackReader.ReadString('\n')
-		if err == io.EOF {
-			break
-		}
-		if err != nil {
-			// We're reading using bytes.NewReader which should never fail.
-			panic("bufio.NewReader failed on a fixed string")
-		}
+type stackParser struct {
+	scan   *scanner
+	stacks []Stack
+	errors []error
+}
+
+func newStackParser(r io.Reader) *stackParser {
+	return &stackParser{
+		scan: newScanner(r),
+	}
+}
+
+func (p *stackParser) Parse() ([]Stack, error) {
+	for p.scan.Scan() {
+		line := p.scan.Text()
 
 		// If we see the goroutine header, start a new stack.
-		isFirstLine := false
 		if strings.HasPrefix(line, "goroutine ") {
-			// flush any previous stack
-			if curStack != nil {
-				stacks = append(stacks, *curStack)
-			}
-			id, goState := parseGoStackHeader(line)
-			curStack = &Stack{
-				id:        id,
-				state:     goState,
-				fullStack: &bytes.Buffer{},
+			stack, err := p.parseStack(line)
+			if err != nil {
+				p.errors = append(p.errors, err)
```

Comment:
does it make sense to continue scanning once we hit an error? especially since any error means we don't use the results

Reply:
The only difference is that it'll report all the errors in all the traces instead of stopping at the first one.
I don't feel strongly about doing it this way. Do you think it's better not to?

Reply:
Only reason not to do this is if a failure here can cause subsequent parses to fail. However, since this method only looks for the `goroutine ` prefix, which wouldn't be affected by where we stopped parsing, I think this is fine.